File: java/Dson-Core/src/main/java/cn/wjybxx/dson/DsonLong4.java
/*
 * Copyright 2026 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson;

import cn.wjybxx.dson.types.Long4;

import javax.annotation.Nonnull;
import java.util.Objects;

/** @author wjybxx */
public final class DsonLong4 extends DsonValue {

    public static final DsonLong4 EMPTY = new DsonLong4(Long4.EMPTY);

    private final Long4 value;

    public DsonLong4(Long4 value) {
        this.value = Objects.requireNonNull(value);
    }

    public Long4 getValue() {
        return value;
    }

    @Nonnull
    @Override
    public DsonType getDsonType() {
        return DsonType.LONG4;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        DsonLong4 that = (DsonLong4) o;
        return value.equals(that.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return "DsonLong4{" +
                "value=" + value +
                '}';
    }
}

File: csharp/Wjybxx.Dson.Core/src/DsonLong4.cs
using System;
using Wjybxx.Dson.Types;

namespace Wjybxx.Dson
{
public sealed class DsonLong4 : DsonValue, IEquatable<DsonLong4>
{
    public static readonly DsonLong4 EMPTY = new DsonLong4(default);
    private readonly Long4 _value;

    public DsonLong4(Long4 value) {
        _value = value;
    }

    public override DsonType DsonType => DsonType.Long4;
    public Long4 Value => _value;

    public bool Equals(DsonLong4? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return _value.Equals(other._value);
    }

    public override bool Equals(object? obj) {
        return obj is DsonLong4 other && Equals(other);
    }

    public override int GetHashCode() {
        return _value.GetHashCode();
    }

    public static bool operator ==(DsonLong4? left, DsonLong4? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonLong4? left, DsonLong4? right) {
        return !Equals(left, right);
    }

    public override string ToString() {
        return $"{nameof(DsonType)}: {DsonType}, {nameof(_value)}: {_value}";
    }
}
}